beacon-chain/sync/subscriber.go

```diff
@@ -412,6 +438,155 @@ func (s *Service) subscribeAggregatorSubnet(
+			select {
+			case <-s.ctx.Done():
+				ticker.Done()
+				return
+			case <-ticker.C():
+				if s.chainStarted.IsSet() && s.cfg.InitialSync.Syncing() {
+					continue
+				}
+				valid, err := isDigestValid(digest, genesis, genRoot)
+				if err != nil {
+					log.Error(err)
+					continue
+				}
+				if !valid {
+					log.Warnf("Sync subnets with digest %#x are no longer valid, unsubscribing from all of them.", digest)
+					// Unsubscribes from all our current subnets.
+					for i := uint64(0); i < params.BeaconConfig().SyncCommitteeSubnetCount; i++ {
+						fullTopic := fmt.Sprintf(topic, digest, i) + s.cfg.P2P.Encoding().ProtocolSuffix()
+						s.unSubscribeFromTopic(fullTopic)
+					}
+					ticker.Done()
+					return
+				}
+				// Check every slot that there are enough peers
+				for i := uint64(0); i < params.BeaconConfig().SyncCommitteeSubnetCount; i++ {
+					if !s.validPeersExist(s.addDigestAndIndexToTopic(topic, digest, i)) {
+						log.Debugf("No peers found subscribed to sync gossip subnet with "+
+							"committee index %d. Searching network for peers subscribed to the subnet.", i)
+						_, err := s.cfg.P2P.FindPeersWithSubnet(
+							s.ctx,
+							s.addDigestAndIndexToTopic(topic, digest, i),
+							i,
+							params.BeaconNetworkConfig().MinimumPeersInSubnet,
+						)
+						if err != nil {
+							log.WithError(err).Debug("Could not search for peers")
+							return
+						}
+					}
+				}
+			}
+		}
+	}()
+}
+
+// subscribe to a dynamically changing list of subnets. This method expects a fmt compatible
+// string for the topic name and the list of subnets for subscribed topics that should be
+// maintained.
+func (s *Service) subscribeDynamicWithSyncSubnets(
+	topicFormat string,
+	validate pubsub.ValidatorEx,
+	handle subHandler,
+	digest [4]byte,
+) {
+	genRoot := s.cfg.Chain.GenesisValidatorRoot()
+	_, e, err := p2putils.RetrieveForkDataFromDigest(digest, genRoot[:])
+	if err != nil {
+		panic(err)
+	}
+	base := p2p.GossipTopicMappings(topicFormat, e)
+	if base == nil {
+		panic(fmt.Sprintf("%s is not mapped to any message in GossipTopicMappings", topicFormat))
+	}
+	subscriptions := make(map[uint64]*pubsub.Subscription, params.BeaconConfig().SyncCommitteeSubnetCount)
+	genesis := s.cfg.Chain.GenesisTime()
+	ticker := slotutil.NewSlotTicker(genesis, params.BeaconConfig().SecondsPerSlot)
+
+	go func() {
+		for {
+			select {
+			case <-s.ctx.Done():
+				ticker.Done()
+				return
+			case currentSlot := <-ticker.C():
+				if s.chainStarted.IsSet() && s.cfg.InitialSync.Syncing() {
+					continue
+				}
+				valid, err := isDigestValid(digest, genesis, genRoot)
+				if err != nil {
+					log.Error(err)
+					continue
+				}
+				if !valid {
+					log.Warnf("Sync subnets with digest %#x are no longer valid, unsubscribing from all of them.", digest)
+					// Unsubscribes from all our current subnets.
+					s.reValidateSubscriptions(subscriptions, []uint64{}, topicFormat, digest)
+					ticker.Done()
+					return
+				}
```

Comment:
Im don't see why we need to put this in the for select loop. `digest, genesis, genRoot` doesn't change
Wouldn't it make more sense to put this validation outside of go routine and short circuit the exit early

Reply:
so `isDigestValid` checks if the provided digest is our current digest, if it is not we would exit this method. This is to handle unsubscribing at fork boundaries.